### DISCUSSION TITLE
Custom Animation Class

### DISCUSSION THREAD

**Discussion by AhmedMajedKaz** [ManimCommunity/manim#1669] [Q&A] [answered]
I have two question, Can you please answer it if you know. First: I search in "https://github.com/3b1b/videos" and found some classes that are inherited from Animation and Transform Class, my question is how to make my custom animation class ?, I don't think this is mentioned in the documentation of Manim. Second: how transform animation work? I search in Transform Class, but I don't understand how it work because I learn python recently, And I want to know how transform and interpolate between two shapes happens. and thanks.

**Comment by sahilmakhijani** [comment, marked as answer]
```python
class RotationAnimation(Animation):
    def __init__(self, mobject, angle, **kwargs):
        self.angle = angle
        super().__init__(mobject, **kwargs)
    
    def interpolate_mobject(self, alpha):
        self.mobject.set_points(self.starting_mobject.get_points())
        self.mobject.rotate(alpha * self.angle)

class RotationScene(Scene):
    def construct(self):
        self.play(
            RotationAnimation(Square(), 45 * DEGREES),
            rate_func=linear,
            run_time=1
        )
```

Check this example for Custom Animation.

While defining custom animation, you should know what alpha is. Alpha ranges from 0 to 1 throughout the animation. For rotation, Now, when animation starts I want object to be 0 degrees tilted from starting object and at the end it should be 45 degrees tilted. So, if I tilt alpha * 45 Degrees, then it should work properly.

If the frame rate is 60 and runtime is 1 second, so interpolate_mobject will be called 60 time per second. At first alpha will be 0, so 0 degress tilt. After half second, alpha will be 0.5, so you will tilt 22.5 degrees. At last alpha will be 1, so you will tilt 45 degrees. That's why it would be a smooth animation from 0 degrees to 45 degrees tilt.

**Reply by WampyCakes** [reply]
Your paragraph explanations could be a helpful example in the docs on how animations actually work in Manim. Perhaps a bit of editing would be needed so that it fits in the context of the docs, but this seems good 💯

**Reply by sahilmakhijani** [reply]
That's Great!!
Feel free to use this example in docs. Let me know if you want anything from my side 😄
Thanks

**Reply by WampyCakes** [reply]
Would you mind making a PR to add it?

**Reply by sahilmakhijani** [reply]
Is this example okay for documentation? If yes, then I'll write one or two paragraphs explaining this example. Then, please guide me on where to put the example (Animation class docstring?).

**Reply by sahilmakhijani** [reply]
Also, please let me know the page where this example will be rendered. Then. I'll write the explanation that fits the context of that page.

**Reply by WampyCakes** [reply]
I think [here](https://docs.manim.community/en/stable/tutorials/building_blocks.html#animations) would work. The internals of Animation are vaguely described by the paragraphs there, but I think your paragraphs + example make it a lot clearer.

**Reply by sahilmakhijani** [reply]
Should I add the **Custom Animation** section between **Animations** and **Animation Methods** sections in https://github.com/ManimCommunity/manim/blob/main/docs/source/tutorials/building_blocks.rst file?

**Reply by WampyCakes** [reply]
I would say either put it at the end of the animation section (so right before using the coordinates of a mobject) or beef up the section called Animations with your explanation and example. Do whichever your discernment tells you is better (which might be decided based on however you edit your explanation)

**Reply by WampyCakes** [reply]
And yes, that file

**Reply by sahilmakhijani** [reply]
Ok thanks, I'll make a PR till next week. 😃

**Reply by sahilmakhijani** [reply]
Is the below example okay for custom animation?

I was thinking to add an example that justifies the need for custom `Animation`. Keeping that in mind, I made this example.

Now, for an explanation, I'll start with "You will often find yourself in a situation where you have to do a smooth transition/animation from one state of an object to another. Manim has lots of built-in animations but if you find yourself in that situation, you could define your custom animation. ... For example, If you have made a bar chart. And, to show it, you have to smoothly increment the bar labels and increse bar heights. ..." (Something like this)

1. Is this example okay?
2. Is code too big for example? (since other examples in docs are very small)
3. Should I add any comments to the following code? (I can't think of any, apart from animation explanation that I'll explain in paragraph)
4. There are two hacks for this code to work: the two `if` conditions in `CreateBarChart.interpolate_mobject`. It is because of problems with Manim's internal workings. I am not able to set the height of `Rectangle` to zero and also, I can't make two-digit `Interger` to one digit. So, what to do?

<details><summary>Code</summary>

```python
from manim import *

class Bar(VGroup):
    def __init__(
        self,
        value,
        range=(0, 100),
        bar_max_height=4,
        bar_rect_config=dict(
            stroke_color=WHITE,
            fill_color=BLUE,
            fill_opacity=1.0,
            width=2
        ),
        **kwargs
    ):
        super().__init__(**kwargs)
        self.value = value
        self.range = range
        self.bar_height = bar_max_height * (value - range[0]) / (range[1] - range[0])
        self.bar_rect = Rectangle(height=self.bar_height, **bar_rect_config)
        self.bar_label = Integer(value)
        self.add(self.bar_rect, self.bar_label)
        self.arrange(DOWN)


class BarChart(VGroup):
    def __init__(
        self,
        values,
        range=(0, 100),
        bar_max_height=4,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.values = values
        self.range = range
        self.bars = [
            Bar(value, range, bar_max_height, **kwargs)
            for value in values
        ]
        self.add(*self.bars)
        self.arrange(aligned_edge=DOWN)


class CreateBarChart(Animation):
    def __init__(self, mobject: BarChart, **kwargs) -> None:
        super().__init__(mobject, ** kwargs)

    def interpolate_mobject(self, alpha: float) -> None:
        assert(isinstance(self.mobject, BarChart))

        for bar in self.mobject.bars:
            height = alpha * bar.bar_height
            value = bar.range[0] + (alpha * (bar.value - bar.range[0]))

            if height > 0:
                bar.bar_rect.stretch_to_fit_height(height, about_edge=DOWN)
            if value >= 10:
                bar.bar_label.set_value(int(value))


class ExplainBarChart(Scene):
    values = [73, 81, 56, 63, 99, 60]

    def construct(self):
        bar_chart = BarChart(self.values)
        self.play(
            CreateBarChart(bar_chart),
            run_time=5,
            rate_func=linear
        )
        self.wait()
```

</details>

<details><summary>Output</summary>

https://user-images.githubusercontent.com/49190295/124136589-8bcda780-daa2-11eb-85df-76497299d428.mp4

</details>

**Reply by WampyCakes** [reply]
I really like this example, but I think you may be right. It's best if we can avoid creating our own mobject for the example and instead we focus on creating a custom animation class. I think that covers 1 and 2.
3. Considering it may be a bit too much of an explanation and need changed, no.
4. I don't know exactly what the problem is anyhow 😅 

Regarding any copyediting, we'll take care of that in the PR. So once you have a legible example, we can fine tune the wording of it :) Thanks for your work thus far! 

I think the introduction looks good btw

**Reply by sahilmakhijani** [reply]
I can't think of an example where there is no custom Mobject and it also justifies the need of custom animation. So, should I just add the rotation example that I've answered on this thread? Any other ideas for custom animation examples?

**Reply by sahilmakhijani** [reply]
`Rotate` is manim's inbuilt animation, that's why I changed the example for docs

**Reply by sahilmakhijani** [reply]
I guess this thred is not the best place to discuss this. Should I open a seperate issue or draft pr for this?

**Reply by WampyCakes** [reply]
I don't think it needs to be justified. You can use the example of something already built into manim, like rotate, to explain how it works so that when someone has their own thing they want to animate, they can see how to do it. They'll already have their own justification at that point. That's all this example really needs to convey I think. 

A draft PR works

**Comment by sahilmakhijani** [comment]
For the second question, transform animation works by calling interpolate on the corresponding type of `Mobject`, which can be `VMobject`, `ImageMobject`, etc. Interpolate method of each type of mobject takes care of how to *interpolate* between two mobjects and gives a transforming effect.
